File: LeetCode/P543_Easy_Diameter_of_Binary_Tree/main.py
from typing import Optional
# Problem: https://leetcode.com/problems/diameter-of-binary-tree/


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        hcache = {}

        def height(root):
            if not root:
                return 0
            if root in hcache:
                return hcache[root]
            hcache[root] = 1 + max(height(root.left), height(root.right))
            return hcache[root]

        def diameter(root):
            if not root:
                return 0
            return max(diameter(root.left), diameter(root.right), 1 + height(root.left) + height(root.right))

        return diameter(root) - 1

    def diameterOfBinaryTree2(self, root: Optional[TreeNode]) -> int:
        def diameter(node, res):
            # Base case
            if not node:
                return 0

            # rec calc left and right diam
            left = diameter(node.left, res)
            right = diameter(node.right, res)

            # update max diameter found
            res[0] = max(res[0], left+right)

            return max(left, right) + 1

        res = [0]
        diameter(root, res)
        return res[0]

File: LeetCode/P543_Easy_Diameter_of_Binary_Tree/main.cpp
#include <unordered_map>
// Problem: https://leetcode.com/problems/diameter-of-binary-tree/

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  // soln 1
  std::unordered_map<TreeNode *, int> heights;

  int diameterOfBinaryTree(TreeNode *root) {
    if (!root)
      return 0;

    return std::max(height(root->left) + height(root->right) + 1,
                    std::max(diameterOfBinaryTree(root->left),
                             diameterOfBinaryTree(root->right))) -
           1;
  }

  int height(TreeNode *root) {
    if (!root)
      return 0;
    if (!heights.contains(root))
      heights[root] = 1 + std::max(height(root->left), height(root->right));

    return heights[root];
  }

  // soln 2
  int diameter(TreeNode *root, int res[]) {
    if (!root)
      return 0;

    int left = diameter(root->left, res);
    int right = diameter(root->right, res);

    res[0] = std::max(res[0], left + right);
    return std::max(left, right) + 1;
  }

  int diameterOfBinaryTree2(TreeNode *root) {
    int res[] = {0};
    diameter(root, res);
    return res[0];
  }
};